File: depmesh_ai/sources/maven.py
"""Maven Central source (repo1.maven.org).

Package names use the ``groupId:artifactId`` form, as in the original
DepMesh Maven prototype. This talks to the repository directly instead of
spawning a local Maven process the way depmesh-resolver did — version
lists come from maven-metadata.xml, release dates from the directory
listing, and the license from the latest POM.
"""
from __future__ import annotations

import datetime as dt
import re
import xml.etree.ElementTree as ET

from ..model import Ecosystem, PackageFacts, ReleaseRef
from .http import NotFound, SourceUnavailable, get_text

REPO = "https://repo1.maven.org/maven2"

# Directory listing lines look like:
#   <a href="3.12.0/" title="3.12.0/">3.12.0/</a>  2021-03-01 07:38  -
_LISTING_ROW = re.compile(
    r'href="(?P<version>[^"/]+)/"[^>]*>[^<]+</a>\s+(?P<date>\d{4}-\d{2}-\d{2})'
)


def _coordinates(name: str) -> tuple[str, str]:
    if ":" not in name:
        raise ValueError(
            f"maven package names must be 'groupId:artifactId', got {name!r}"
        )
    group, artifact = name.split(":", 1)
    return group, artifact


def _listing_dates(base_url: str) -> dict[str, dt.date]:
    try:
        listing = get_text(f"{base_url}/")
    except (NotFound, SourceUnavailable):
        return {}
    dates = {}
    for match in _LISTING_ROW.finditer(listing):
        try:
            dates[match.group("version")] = dt.date.fromisoformat(match.group("date"))
        except ValueError:
            pass
    return dates


def _pom_license(base_url: str, artifact: str, version: str) -> str | None:
    try:
        pom = get_text(f"{base_url}/{version}/{artifact}-{version}.pom")
    except (NotFound, SourceUnavailable):
        return None
    # Strip the default-namespace prefix so find() paths stay simple.
    pom = re.sub(r'xmlns="[^"]+"', "", pom, count=1)
    try:
        root = ET.fromstring(pom)
    except ET.ParseError:
        return None
    node = root.find("./licenses/license/name")
    return node.text.strip() if node is not None and node.text else None


def fetch(name: str) -> PackageFacts:
    group, artifact = _coordinates(name)
    facts = PackageFacts(Ecosystem.MAVEN, name)
    base_url = f"{REPO}/{group.replace('.', '/')}/{artifact}"

    try:
        metadata = get_text(f"{base_url}/maven-metadata.xml")
    except NotFound:
        facts.exists = False
        return facts

    facts.exists = True
    try:
        root = ET.fromstring(metadata)
    except ET.ParseError as err:
        raise SourceUnavailable(f"unparseable maven-metadata.xml for {name}") from err

    versioning = root.find("versioning")
    versions: list[str] = []
    if versioning is not None:
        latest = versioning.findtext("release") or versioning.findtext("latest")
        facts.latest_version = latest
        versions = [
            v.text for v in versioning.findall("./versions/version") if v.text
        ]

    dates = _listing_dates(base_url)
    if not dates:
        facts.degraded.append("maven directory listing (release dates unavailable)")
    releases = [ReleaseRef(v, dates.get(v)) for v in versions]
    releases.sort(key=lambda r: (r.release_date or dt.date.min), reverse=True)
    facts.releases = releases

    if facts.latest_version:
        facts.license = _pom_license(base_url, artifact, facts.latest_version)
    return facts

File: internal/sources/maven.go
package sources

import (
	"encoding/xml"
	"errors"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/jhberges/depmesh-ai/internal/model"
)

// Maven Central source (repo1.maven.org). Package names use the
// "groupId:artifactId" form, as in the original DepMesh Maven prototype.
// This talks to the repository directly instead of spawning a local Maven
// process the way depmesh-resolver did — version lists come from
// maven-metadata.xml, release dates from the directory listing, and the
// license from the latest POM.
const mavenRepo = "https://repo1.maven.org/maven2"

// Directory listing rows look like:
//
//	<a href="3.12.0/" title="3.12.0/">3.12.0/</a>  2021-03-01 07:38  -
var listingRow = regexp.MustCompile(`href="([^"/]+)/"[^>]*>[^<]+</a>\s+(\d{4}-\d{2}-\d{2})`)

type mavenMetadata struct {
	Versioning struct {
		Latest   string   `xml:"latest"`
		Release  string   `xml:"release"`
		Versions []string `xml:"versions>version"`
	} `xml:"versioning"`
}

type mavenPOM struct {
	Licenses []struct {
		Name string `xml:"name"`
	} `xml:"licenses>license"`
}

func fetchMaven(name, version string) (*model.PackageFacts, error) {
	group, artifact, ok := strings.Cut(name, ":")
	if !ok {
		return nil, fmt.Errorf("maven package names must be 'groupId:artifactId', got %q", name)
	}
	facts := &model.PackageFacts{Ecosystem: model.Maven, Name: name}
	baseURL := mavenRepo + "/" + strings.ReplaceAll(group, ".", "/") + "/" + artifact

	body, err := getText(baseURL + "/maven-metadata.xml")
	if errors.Is(err, ErrNotFound) {
		facts.Exists = model.Bool(false)
		return facts, nil
	}
	if err != nil {
		return nil, err
	}

	facts.Exists = model.Bool(true)
	var metadata mavenMetadata
	if err := xml.Unmarshal([]byte(body), &metadata); err != nil {
		return nil, &UnavailableError{baseURL, fmt.Errorf("unparseable maven-metadata.xml: %w", err)}
	}
	facts.LatestVersion = firstNonEmpty(metadata.Versioning.Release, metadata.Versioning.Latest)

	dates := listingDates(baseURL)
	if len(dates) == 0 {
		facts.Degraded = append(facts.Degraded, "maven directory listing (release dates unavailable)")
	}
	for _, version := range metadata.Versioning.Versions {
		ref := model.ReleaseRef{Version: version}
		if date, ok := dates[version]; ok {
			date := date
			ref.ReleaseDate = &date
		}
		facts.Releases = append(facts.Releases, ref)
	}
	sortNewestFirst(facts.Releases)

	if facts.LatestVersion != "" {
		facts.License = pomLicense(baseURL, artifact, facts.LatestVersion)
	}

	if version != "" {
		facts.Requested = mavenVersionFacts(facts, baseURL, artifact, version, metadata.Versioning.Versions)
	}
	return facts, nil
}

// mavenVersionFacts resolves a requested version against maven-metadata.xml,
// and on a miss confirms against the POM for that exact coordinate — metadata
// can lag behind what is actually published, so its silence is not proof.
//
// The requested version's license is free: pomLicense already takes a version,
// and this is simply the same single request pointed at a different coordinate
// instead of always at the latest one.
func mavenVersionFacts(facts *model.PackageFacts, baseURL, artifact, version string, published []string) *model.VersionFacts {
	resolved := ""
	for _, candidate := range published {
		if candidate == version {
			resolved = candidate
			break
		}
	}
	vf := requestedVersion(facts, version, resolved, func() (string, error) {
		if _, err := getText(pomURL(baseURL, artifact, version)); err != nil {
			return "", err
		}
		return version, nil
	})
	if vf.Resolved != "" {
		vf.License = pomLicense(baseURL, artifact, vf.Resolved)
	}
	return vf
}

func listingDates(baseURL string) map[string]time.Time {
	listing, err := getText(baseURL + "/")
	if err != nil {
		return nil
	}
	dates := map[string]time.Time{}
	for _, match := range listingRow.FindAllStringSubmatch(listing, -1) {
		if t, err := time.Parse("2006-01-02", match[2]); err == nil {
			dates[match[1]] = t
		}
	}
	return dates
}

func pomURL(baseURL, artifact, version string) string {
	return fmt.Sprintf("%s/%s/%s-%s.pom", baseURL, version, artifact, version)
}

func pomLicense(baseURL, artifact, version string) string {
	body, err := getText(pomURL(baseURL, artifact, version))
	if err != nil {
		return ""
	}
	var pom mavenPOM
	if err := xml.Unmarshal([]byte(body), &pom); err != nil {
		return ""
	}
	if len(pom.Licenses) > 0 {
		return strings.TrimSpace(pom.Licenses[0].Name)
	}
	return ""
}
